solidity/contracts/isms/hook/OPL2ToL1CcipReadIsm.sol

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import {Message} from "../../libs/Message.sol";
+import {TokenMessage} from "../../token/libs/TokenRouter.sol";
+import {TypeCasts} from "../../libs/TypeCasts.sol";
+import {IMailbox} from "../../interfaces/IMailbox.sol";
+import {OPL2ToL1Withdrawal} from "../../libs/OPL2ToL1Withdrawal.sol";
+import {AbstractCcipReadIsm} from "../ccip-read/AbstractCcipReadIsm.sol";
+import {ICcipReadIsm} from "../../interfaces/isms/ICcipReadIsm.sol";
+import {IMessageRecipient} from "../../interfaces/IMessageRecipient.sol";
+import {IOptimismPortal} from "../../interfaces/optimism/IOptimismPortal.sol";
+import {IInterchainSecurityModule, ISpecifiesInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+
+/**
+ * @notice Prove and finalize a OP stack withdrawal on L1
+ * @dev Proving and finalizing had been merged into a single
+ * ISM because OP Stack expects the prover and the finalizer to
+ * be the same caller
+ */
+contract OPL2ToL1CcipReadIsm is
+    AbstractCcipReadIsm,
+    IMessageRecipient,
+    ISpecifiesInterchainSecurityModule
+{
+    using Message for bytes;
+    using TypeCasts for address;
+
+    // CCIP-read gateways URLs
+    string[] public urls;
+    // mailbox on L1
+    IMailbox mailbox;
```

Comment:
this should be `immutable` to avoid incurring `SLOAD` costs
there is also no mutatot fn included